## Cleaning with pandas - askscience_submissions edition

First, let's import pandas and open `../data/askscience_submissions.csv` we previously downloaded for homework as a dataframe. Then, let's look at the data!

In [1]:
import pandas as pd

In [2]:
science_data = pd.read_csv('../data/askscience_submissions.csv')

/var/folders/8c/d75q8fq50n1550n3gfm_93hc0000gn/T/ipykernel_65749/3501743720.py:1: DtypeWarning: Columns (0: approved_at_utc, 1: archived, 2: author_cakeday, 3: banned_at_utc, 4: brand_safe, 5: can_gild, 6: can_mod_post, 7: contest_mode, 8: created, 9: created_utc, 10: crosspost_parent, 11: crosspost_parent_list, 12: distinguished, 13: hidden, 14: hide_score, 15: is_crosspostable, 16: is_reddit_media_domain, 17: is_self, 18: is_video, 19: locked, 20: name, 21: over_18, 22: parent_whitelist_status, 23: pinned, 24: post_hint, 25: preview, 26: quarantine, 27: saved, 28: spoiler, 29: stickied, 30: subreddit_type, 31: suggested_sort, 32: whitelist_status) have mixed types. Specify dtype option on import or set low_memory=False.
  science_data = pd.read_csv('../data/askscience_submissions.csv')


In [3]:
science_data.head(2)

,approved_at_utc,archived,author,author_cakeday,author_flair_css_class,author_flair_text,banned_at_utc,brand_safe,can_gild,can_mod_post,...,subreddit_type,suggested_sort,thumbnail,thumbnail_height,thumbnail_width,title,ups,url,view_count,whitelist_status
0,NaN,True,vertexoflife,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,self,NaN,NaN,Why are so many sodium-based products used in ...,6.0,http://www.reddit.com/r/askscience/comments/1u...,NaN,NaN
1,NaN,True,[deleted],NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,default,NaN,NaN,Why do almost all living organisms have a long...,1.0,http://www.reddit.com/r/askscience/comments/1u...,NaN,NaN


In [4]:
len(science_data)

618576

#### Drop rows with missing values

Sometimes data can be incomplete. We can drop rows with incomplete data by using `.dropna()` and specifying what `subset` of columns we want to scrutinize for missing data. 

Let's remove rows that have `NaN` or a null value, in the `title` or `ups` column.

In [5]:
science_no_or_na_title = science_data.dropna(subset=['title', 'ups'])

If we want to drop rows that have `NaN` or a null value, in the `title` _and_ `ups` column.

In [6]:
science_no_na_title = science_data.dropna(subset=['title', 'ups'], how="all")

In [7]:
science_no_na_title.head(2)

,approved_at_utc,archived,author,author_cakeday,author_flair_css_class,author_flair_text,banned_at_utc,brand_safe,can_gild,can_mod_post,...,subreddit_type,suggested_sort,thumbnail,thumbnail_height,thumbnail_width,title,ups,url,view_count,whitelist_status
0,NaN,True,vertexoflife,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,self,NaN,NaN,Why are so many sodium-based products used in ...,6.0,http://www.reddit.com/r/askscience/comments/1u...,NaN,NaN
1,NaN,True,[deleted],NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,default,NaN,NaN,Why do almost all living organisms have a long...,1.0,http://www.reddit.com/r/askscience/comments/1u...,NaN,NaN


Compare the lengths of the original dataframe and the dataframe without the missing title and ups values using `len()`

In [8]:
print(len(science_data))
print(len(science_no_na_title))

618576
618558


#### Dropping rows based on conditions

Get all of the rows whose `ups` is greater than 99.

In [9]:
science_data[science_data['ups'] > 99]

,approved_at_utc,archived,author,author_cakeday,author_flair_css_class,author_flair_text,banned_at_utc,brand_safe,can_gild,can_mod_post,...,subreddit_type,suggested_sort,thumbnail,thumbnail_height,thumbnail_width,title,ups,url,view_count,whitelist_status
58,NaN,True,AmericanDerp,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,self,NaN,NaN,What would happen in the short and long term i...,158.0,http://www.reddit.com/r/askscience/comments/1u...,NaN,NaN
75,NaN,True,mug3n,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,self,NaN,NaN,How does 'move your mouse randomly' for increa...,1574.0,http://www.reddit.com/r/askscience/comments/1u...,NaN,NaN
130,NaN,True,superjames_16,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,default,NaN,NaN,Why can I eat large quantities of fast food in...,110.0,http://www.reddit.com/r/askscience/comments/1u...,NaN,NaN
153,NaN,True,[deleted],NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,default,NaN,NaN,Why is it so difficult to remember dreams?,119.0,http://www.reddit.com/r/askscience/comments/1u...,NaN,NaN
172,NaN,True,bohring1150,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,self,NaN,NaN,Has insomnia been observed in animals other th...,522.0,http://www.reddit.com/r/askscience/comments/1u...,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
477757,NaN,False,MrAcurite,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,self,NaN,NaN,When did Humans first start eating communal me...,221.0,https://www.reddit.com/r/askscience/comments/5...,NaN,NaN
477769,NaN,False,danredux,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,default,NaN,NaN,How much light can fit in one place?,352.0,https://www.reddit.com/r/askscience/comments/5...,NaN,NaN
477792,NaN,False,acrowsmurder,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,self,NaN,NaN,How would we know if Voyager 1 or 2 crashed in...,969.0,https://www.reddit.com/r/askscience/comments/5...,NaN,NaN
477896,NaN,False,jmel79,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,self,NaN,NaN,"If we could drain the ocean, could we breath o...",18789.0,https://www.reddit.com/r/askscience/comments/5...,NaN,NaN


#### Making new columns

Sometimes we need to filter data based on more complicated conditions. We can do that by making a new column where each row is either `True` or `False`, based on criteria that we make up.

Let's do this by making a column called `contains_climate_change` that is `True` or `False` based on whether the phrase "climate change" is in the title.

In [10]:
science_data['contains_climate_change'] = science_data['title'].str.contains('climate change')

Now let's make a variable that only has the rows where `contains_climate_change` is `True`

In [11]:
science_data_climate = science_data[science_data['contains_climate_change'] == True]

In [12]:
len(science_data_climate)

2221

Gutcheck: Is the length of this filtered data 2221? If so, you've done it correctly!

#### Paring down columns

For our variable that only has rows that contain the phrase "climate change" in the title, select only the columns `author`, `title`, `ups`, and `downs`, and then use the `sort_values()` function we touched upon last week to get the name of the author with the highest upvotes ("`ups`").

In [13]:
columns = ['author', 'title', 'ups', 'downs']
science_data_climate = science_data_climate[columns]

In [14]:
science_data_climate.sort_values(by="ups", ascending=False)

,author,title,ups,downs
376781,AmGeophysicalU-AMA,"AGU AMA: I’m Dr. Kim Cobb, and I’m here to tal...",2862.0,0.0
21992,AskScienceModerator,Questions about the climate change debate betw...,1294.0,0.0
267712,ari_daniel,"AskScience AMA Series: I'm Ari Daniel, science...",1277.0,0.0
188908,level10peon,What is the likely range of possibilities for ...,621.0,0.0
345362,iorgfeflkd,Is anthropogenic climate change predicted to m...,195.0,0.0
...,...,...,...,...
615676,Deadwalker123,What exactly is it that humans are doing to ca...,NaN,NaN
616541,RealShigeruMeeyamoto,"Why do ""small"" changes of temperatures caused ...",NaN,NaN
616640,ArbitraryNarrative,"Can the impact of climate change, specifically...",NaN,NaN
616983,PM_ME_YOUR_GOOD_DOGS,How credible of a documentary is 'the great cl...,NaN,NaN


#### Saving data

Let's save that using the `.to_csv()` function!

In [16]:
science_data_climate.to_csv('../output/just_climate.csv')